File: Leetcode/2187 Minimum Time to Complete Trips.py
class Solution:
    def minimumTime(self, times, total_trips):
        min_time = min(times)
        lo, hi = min_time, total_trips * lo
        time = 0
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            if can_make_trips(mid, times, total_trips):
                time = mid
                hi = mid - 1
            else:
                lo = mid + 1
        return time


def can_make_trips(time, times, total_trips):
    trips = sum(map(lambda t: time // t, times))
    return trips >= total_trips

File: Leetcode/2187 Minimum Time to Complete Trips.rs
impl Solution {
    pub fn minimum_time(times: Vec<i32>, total_trips: i32) -> i64 {
        let total_trips = total_trips as i64;
        let min_time = (*times.iter().min().unwrap()) as i64;
        let mut lo = min_time;
        let mut hi = min_time * total_trips;
        while lo <= hi {
            let mid = (lo + hi) / 2;
            if Self::can_make_trips(mid, &times, total_trips) {
                hi = mid - 1
            } else {
                lo = mid + 1
            }
        }
        lo
    }

    fn can_make_trips(time: i64, times: &Vec<i32>, total_trips: i64) -> bool {
        let trips: i64 = times.iter()
                              .map(|t| time / (*t as i64))
                              .sum();
        trips >= total_trips
    }
}